homeassistant/components/airq/__init__.py

```diff
@@ -0,0 +1,96 @@
+"""The air-Q integration.
+
+This file defines the initialisation of the integration, invoked from ConfigFlow.
+Integration setup defined here calls out to the platform setup (see sensors.py).
+"""
+from __future__ import annotations
+
+from datetime import timedelta
+import logging
+
+from aioairq import AirQ
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_IP_ADDRESS, CONF_PASSWORD, Platform
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+PLATFORMS: list[Platform] = [Platform.SENSOR]
+
+
+class AirQCoordinator(DataUpdateCoordinator):
+    """Coordinator is responsible for querying the device at a specified route."""
+
+    config: dict = {}
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        update_interval: timedelta,
+        address: str,
+        passw: str,
+        target_route: str = "average",
+    ) -> None:
+        """Initialise a custom coordinator."""
+        super().__init__(hass, _LOGGER, name=DOMAIN, update_interval=update_interval)
+        self.airq = AirQ(address, passw)
+        self._target_route = target_route
+
+    async def _async_update_data(self) -> dict:
+        """Fetch the data from the device.
+
+        Function is meant as an async closure, or partial(airq.get, target_route)
+        Additionally, the result dictionary is stripped of the errors. Subject to
+        a discussion
+        """
+        data = await self.airq.get(self._target_route)
+        return self.airq.drop_errors_from_data(data)
```

Comment:
Some words about upstream library:
1. I noticed that you are creating a session every time you request. It would be better to create the session once (or get existed) using HA helper and pass into your library. ([example](https://github.com/home-assistant/core/blob/f3e432c9c7735daddd2946773cc6572567821a5e/homeassistant/components/adguard/__init__.py#L48-L57), [example](https://github.com/home-assistant/core/blob/f3e432c9c7735daddd2946773cc6572567821a5e/homeassistant/components/accuweather/__init__.py#L38-L42), [example](https://github.com/home-assistant/core/blob/f3e432c9c7735daddd2946773cc6572567821a5e/homeassistant/components/braviatv/__init__.py#L27-L30))
2. Instead of dropping errors (drop_errors_from_data), it is customary to throw an UpdateFailed exception here ([example](https://github.com/home-assistant/core/blob/f3e432c9c7735daddd2946773cc6572567821a5e/homeassistant/components/braviatv/coordinator.py#L124-L127)). However, aiohttp and some commons exceptions can be caught automatically by the DataUpdateCoordinator if they are not caught at the library level. See [DataUpdateCoordinator](https://github.com/home-assistant/core/blob/f3e432c9c7735daddd2946773cc6572567821a5e/homeassistant/helpers/update_coordinator.py#L214-L219).

Reply:
1. Thanks for the suggestion. It sounds good and I will look into it.
2. Perhaps, `drop_errors_from_data` is a misnomer, as by errors this method means uncertainties and not exceptions. Specifically, it drops uncertainties of the measurement values, which are estimated by the device and returned by `self.airq.get()`. I'll rename the method.

Reply:
Hey @Drafteed, I have refactored the upstream module to accept a session, and made according changes. Thank you for the suggestion!
I also renamed `drop_errors_from_data` method and added a formerly missing docstring.